component/status/status.go

```diff
@@ -0,0 +1,138 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package status // import "go.opentelemetry.io/collector/component/status"
+
+import (
+	"time"
+
+	"go.opentelemetry.io/collector/config"
+)
+
+// EventType represents an enumeration of status event types
+type EventType int
+
+const (
+	// EventOK indicates the producer of a status event is functioning normally
+	EventOK EventType = iota
+	// EventError indicates the producer of a status event is in an error state
+	EventError
+)
+
+// StatusEvent is an event produced by a component to communicate its status to registered listeners.
+// An event can signal that a component is working normally (i.e. Type: status.EventOK), or that it
+// is in an error state (i.e. Type: status.EventError). An error status may optionally include an
+// error object to provide additonal insight to registered listeners.
+type StatusEvent struct {
+	Type        EventType
+	Timestamp   int64
+	ComponentID config.ComponentID
+	Error       error
+}
+
+// StatusEventOption applies options to a StatusEvent
+type StatusEventOption func(*StatusEvent)
+
+// WithError assigns an error object to a StatusEvent. It is optional and should only be applied
+// to a StatusEvent with type status.EventError.
+func WithError(err error) StatusEventOption {
+	return func(o *StatusEvent) {
+		o.Error = err
+	}
+}
+
+// WithTimestamp sets the timestamp, expected in nanos, for a StatusEvent
+func WithTimestamp(nanos int64) StatusEventOption {
+	return func(o *StatusEvent) {
+		o.Timestamp = nanos
+	}
```

Comment:
I got rid of the options and left everything mutable. I would have gone the immutable route, but we need to be able to set the component id via the `hostWrapper`. I also got rid of `NewStatusEvent`. It was useful for initializing the timestamp. I moved that to the `hostWrapper` for now, but it isn't obvious that it gets set there unless you thoroughly study the code. Should I have functions to help construct a StatusEvent even with all fields exported? Would `NewStatusEvent(eventType status.EventType)` and `NewStatusEventWithError(eventType status.EventType, err error)` be helpful or something else?

Reply:
>  I would have gone the immutable route, but we need to be able to set the component id via the hostWrapper

A way around this is to define the status on `service/internal/components` (or another suitable internal package where you also define the host wrapper) and then define an alias for it (`type StatusEvent = component.StatusEvent`) on `component/status`.

Reply:
I generally like this idea and I see how this could allow us to update the component id from the host wrapper, but I sketched out the details of what I think this would require, and I'm not sure we want to go this direction. Since the host wrapper is in `service/internal/components` we would have to define `StatusEvent` in the same package, and alias it in the `service` package to make it publicly available as `service.StatusEvent`. If we wanted to make it available from the component package we'd have to alias it a second time to have `component.StatusEvent`. We would have two aliases, which seems less than ideal. Since the goal would be to make the `StatusEvent` immutable to the outside world, we'd have to do the same for the constructor and the `StatusEventOption`s as well. One last issue, is that I tried to alias a type from `service/internal` to `service` to `component` and found that it creates a circular dependency (e.g. import cycle not allowed). It's possible I'm going about this the wrong way, if so, let me know, but if this is accurate, it seems like this path has some unintended consequences.

As things are currently implemented the `service.ReportStatus` function has a chance to modify a `StatusEvent` before passing it off to listeners (to update component id, for example). The event is passed by value to the listeners, so a change to a field by one listener, will not be seen by other listeners. Given that, I'm not overly concerned with all the fields being public on `StatusEvent`. If others have concerns, let me know. I wish we had a better way to initialize the timestamp when creating an event though.

Reply:
After thinking this through, I think we could make our lives easier by making `StatusEvent` immutable and update it by making a copy with the new component id in the host wrapper. I'll move forward with that approach if there aren't any objections.